Multithreaded Benchmarks

lanice opened the pull request:
Closes #1005.

This adds the parameters `--cores X` and `--clients X` to the benchmarkRunner, with `cores` specifying the number of cores that will be used IF `--scheduler` is active, and `clients` setting the number of simulated clients that schedule queries in parallel.

This also adds a python script, `scripts/benchmark_multithreaded.py`, which allows for benchmarking over multiple numbers of cores and putting the results together into a plot. Currently for the TPC-H benchmark only.

The python script takes the following parameters:
- `EXECUTABLE`: mandatory, the path to the `hyriseBenchmarkTPCH` executable
- `--runs X` OR `--runs-per-core X`: one of the two has to be set. `--runs X` means, no matter the number of cores, we always pass `--runs X` to the Hyrise benchmark. `--runs-per-core X` means we pass `--runs Y` to the Hyrise benchmark, with Y being X times the number of cores of the current benchmark run
- `--max-cores X` OR `--cores X Y Z`: optional. If neither is set, the benchmark will be executed for a range of cores from 1 to the maximum available number of cores. `--cores X Y Z` specifies which core counts exactly should be benchmarked
- `--queries X Y Z`: optional. If not set, all possible TPC-H queries are benchmarked, otherwise the specified
- `--result-dir` and `result-name`: optional, defaults are `results/` and current datetime
- `--verbose`: print debug output
- all other given parameters (like `scale`, `chunk_size`, etc) will be passed on to the Hyrise benchmark

Here an example execution:
```
python3 ./scripts/benchmark_multithreaded.py ./build-release/hyriseBenchmarkTPCH --runs 1000 -v
```
The final plot could look like this:
![plots1](https://user-images.githubusercontent.com/1871704/44757422-3ec96a00-ab2f-11e8-8a1f-4b029b03d73a.png)

And another example, this time a bit more specified:
```
python3 ./scripts/benchmark_multithreaded.py ./build-release/hyriseBenchmarkTPCH --runs 1000 -v --scale 1 --cores 10 20 30 --queries 1 3 6 12
```
Possible output plot:
![plots2](https://user-images.githubusercontent.com/1871704/44757450-6ae4eb00-ab2f-11e8-86e2-8980846442bd.png)

mrks commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,190 @@
+    parser.add_argument('--chunk-size', action='store', type=int, metavar='S', help='Specify maximum chunk size (default: Maximum available)')
+    parser.add_argument('--result-dir', action='store', type=str, metavar='DIR', default='results', help='Directory where the result folder will be stored (default: \'results/\')')
+    parser.add_argument('--result-name', action='store', type=str, metavar='NAME', help='Directory where the actual results will be stored (default: current datetime)')
+    parser.add_argument('--build-dir', action='store', type=str, metavar='DIR', required=True, help='Directory that contains the hyriseBenchmarkTPCH executable')
```
Why do you take the build directory instead of the binary? How about this syntax, which allows us to pass any type of arguments to the binary without explicitly listing them in this script:

`benchmark_multithreaded.py [--script-options=foo] ./build-release/hyriseBenchmarkTPCH [--hyrise_options]`

mrks commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,190 @@
+
+        if core_count == 0:
+            use_scheduler = 'false'
+            number_of_runs = args.runs_per_core if args.runs_per_core else args.fixed_runs
```
Thinking about this again, I still think this is more complicated than it has to be. What if we allow `--runs=0` on the Hyrise side which means "run until time is up". Wouldn't that solve the problem here?

lanice replied:
Not sure what problem you mean here. If I specify `--runs=0` (and if that's allowed on the Hyrise side), then that will be used and passed to Hyrise. But if I specify `--runs-per-core=X`, then that instead will be used to determine the number of max query runs for this benchmark run.

mrks commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -103,21 +104,44 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
-  BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
-  while (state.keep_running()) {
-    std::shuffle(mutable_named_queries.begin(), mutable_named_queries.end(), random_generator);
+  if (_config.parallel_execution) {
```
If `--parallel_clients` took an integer, we wouldn't need this `if`, correct?

mrks commented on src/benchmarklib/benchmark_utils.cpp:
```diff
@@ -140,7 +143,21 @@ BenchmarkConfig CLIConfigParser::parse_basic_options_json_config(const nlohmann:
 
   const auto enable_scheduler = json_config.value("scheduler", default_config.enable_scheduler);
-  out << "- Running in " + std::string(enable_scheduler ? "multi" : "single") + "-threaded mode" << std::endl;
+  const auto available_cores = json_config.value("cores", default_config.available_cores);
```
This should verify that the scheduler is enabled

lanice commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,190 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-s', '--scale', action='store', type=float, metavar='S', default=0.1, help='TPC-H scale factor (default: 0.1)')
+    parser.add_argument('-v', '--verbose', action='store_true', help='Print log messages')
+    parser.add_argument('-q', '--queries', action='store', type=int, metavar='Q', nargs='+', help='Specify the TPC-H queries that will be benchmarked')
```
To your first question: No, `--queries 1 3 5` is the way python argparse handles "list" parameters, comma-separated doesn't work.

lanice replied:
To your second question, that is unfortunately a `cxxopt` problem, which basically cannot handle "list" inputs like that. You'd have to use the `--queries` parameter for each query you want to add:
```
./build/hyriseBenchmarkTPCH --queries 1 --queries 2 -s 0.001
```
That's why I added the shortcut `-q` for `--queries` there, because it makes life much better when you instead can just do:
```
./build/hyriseBenchmarkTPCH -q 1 -q 2 -s 0.001
```

mrks replied:
Maybe accept a string in both cases and split it manually?

lanice commented:
@mrks As far as the cpp code goes, I addressed the review comments and what we talked about in person, except the one with the cxxopts parameter to specify queries.
I will adjust the python script as well, but if you want, you can already have a look at the hyrise code.

mrks commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
+  auto tasks = std::vector<std::shared_ptr<AbstractTask>>{};
+  auto state = BenchmarkState{_config.max_num_query_runs, _config.max_duration};
+
+  while (state.keep_running(finished_query_set_runs.load(std::memory_order_relaxed))) {
```
I don't understand this line

lanice replied:
Which part?
1) `state.keep_running()` now takes a parameter of how many query (sets) are already finished (will document that in the BenchmarkState declaration)
2) `.load()` is to get the actual value from std::atomic
3) `std::memory_order_relaxed` tells `load` that there are no synchronization or ordering constraints imposed on other reads or writes, only this operation's atomicity is guaranteed. Which, according to [this](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering), should be fine here. However after thinking more about it now, I will reevaluate the memory_ordering uses in the other cases as well.

mrks replied:
The first. Even with the documentation, I think that the signature of that method is everything but easy to understand. Why do I have to pass a number into the state to find out if the benchmark should continue? Can't we do the check outside of `keep_running`, i.e.

`while(state.keep_running() && ...`?

lanice replied:
Yes we can. This would leave the BenchmarkState as a timer basically.

mrks commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
+                              &finished_query_set_runs, &finished_queries_total, &result_mutex, this]() {
+          const auto duration = std::chrono::steady_clock::now() - query_run_begin;
+          finished_queries_total++;
+          if (finished_queries_total.load(std::memory_order_relaxed) % number_of_queries_per_set == 0) {
```
Why not reuse the return value of `++`?

lanice replied:
Yeah that... makes sense...

lanice commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
-      auto& query_benchmark_result = _query_results_by_query_name[named_query.first];
-      query_benchmark_result.duration += query_benchmark_end - query_benchmark_begin;
-      query_benchmark_result.num_iterations++;
+  const auto number_of_queries_per_set = mutable_named_queries.size();
```
With `--mode PermutedQuerySets` we are running on sets of queries, not individual queries. So "set" is one (shuffled) batch of the queries we want to run.

mrks replied:
Got it. I think I find "sets" a bit misleading, as there is only one set of queries that is executed in a random order, isn't there?

lanice replied:
Ah I see what you mean. So basically rename the mode PermutedQuerySets -> PermutedQuerySet ?

And actually, `number_of_queries_per_set` might as well be just `number_of_queries`, which is what that is.

mrks replied:
Personally, I would find that easier to understand

lanice commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -103,45 +104,158 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
+  }
+}
+
+std::vector<std::shared_ptr<AbstractTask>> BenchmarkRunner::_schedule_or_execute_query(
```
The only reason for that is that for Queries that require execution (like TPC-H 15), we can't call get_tasks() on the pipeline to schedule each task. So with that, these queries don't break when we don't use the scheduler.

mrks replied:
Ok, maybe leave a comment for that.

Bouncner commented:
Any changes to the examples you provided in the PR description?

lanice commented:
I haven't updated the python script yet, so the examples won't work right now.

Bensk1 commented:
What's the status on that?

lanice commented:
> What's the status on that?

The PR is done, I'm just waiting for a review.

mrks commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,205 @@
+# To prevent _tkinter.TclError: https://stackoverflow.com/a/37605654
```
add shebang, set `chmod +x` for this file

mrks commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,205 @@
+import json
+
+from datetime import datetime
+
```
```
Markus.Dreseler@nemea:~/hyrise/build-release$ python3 ../scripts/benchmark_multithreaded.py --runs 1000 -q 1 2 ./hyriseBenchmarkTPCH
usage: benchmark_multithreaded.py [-h] [-v] [-q Q [Q ...]] [--result-dir DIR]
                                  [--result-name NAME] [-c C [C ...] |
                                  --max-cores C] (-r N | --runs-per-core N)
                                  EXECUTABLE
benchmark_multithreaded.py: error: argument -q/--queries: invalid int value: './hyriseBenchmarkTPCH'
```

Can we do anything about this?

lanice replied:
We could make the executable a named argument `-e, --executable` instead of a positional argument?

mrks replied:
hmpf. That would mean that if we want to pass parameters to the executable, we would have to enclose it in quotes, right?

And what if, instead of using Python conventions, we use the same query parameter format that Hyrise uses?

mrks replied:
hmpf. That would mean that if we want to pass parameters to the executable, we would have to enclose it in quotes, right?

And what if, instead of using Python conventions, we use the same query parameter format that Hyrise uses?

mrks replied:
Sounds good.

mrks commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,205 @@
+    parser.add_argument('--result-dir', action='store', type=str, metavar='DIR', default='results', help='Directory where the result folder will be stored (default: \'results/\')')
+    parser.add_argument('--result-name', action='store', type=str, metavar='NAME', help='Directory where the actual results will be stored (default: current datetime)')
+    parser.add_argument('executable', action='store', type=str, metavar='EXECUTABLE', help='hyriseBenchmarkTPCH executable')
+    parser.add_argument('--clients', action='store', type=int, metavar='C', default=10, help='The number of clients that schedule queries in parallel')
```
This doesn't show up in the help

lanice replied:
The help for `--clients`? For me it does:

```
lanice:opossumdb (leander/1005_multithreaded_benchmarks) $ ./scripts/benchmark_multithreaded.py --help
usage: benchmark_multithreaded.py [-h] [-v] [-q Q [Q ...]] [--result-dir DIR]
                                  [--result-name NAME] [--clients C]
                                  [-c C [C ...] | --max-cores C]
                                  (-r N | --runs-per-core N)
                                  EXECUTABLE

positional arguments:
  EXECUTABLE            hyriseBenchmarkTPCH executable

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose         Print log messages
  -q Q [Q ...], --queries Q [Q ...]
                        Specify the TPC-H queries that will be benchmarked
  --result-dir DIR      Directory where the result folder will be stored
                        (default: 'results/')
  --result-name NAME    Directory where the actual results will be stored
                        (default: current datetime)
  --clients C           The number of clients that schedule queries in
                        parallel
  -c C [C ...], --cores C [C ...]
                        List of cores to be used for the benchmarks
  --max-cores C         Number of cores up to which the benchmarks will be
                        executed
  -r N, --runs N        Fixed number of (max) runs each query is executed
  --runs-per-core N     Number of (max) runs per core each query is executed
```

mrks replied:
Maybe I am blind. Or I was on the wrong branch. Or... some other lame excuse.

mrks commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,205 @@
+    benchmark_run = 0
+    for core_count in core_counts:
+        benchmark_run += 1
+        verbose_print(args.verbose, 'Starting benchmark run ' + str(benchmark_run) + ' of ' + str(len(core_counts)) + ' ...')
```
I would leave out "run" here. In the context of benchmarks, I understand it as if one benchmark was executed multiple times.

mrks commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,205 @@
+
+    return result_dir
+
+def plot(args, result_dir):
```
Can we get the NUMA boundaries in?

lanice replied:
Yep, I added the number of cores per NUMA node to the result JSON output, and with that the python script can print the boundaries (see updated image below).

mrks commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,205 @@
+
+    return result_dir
+
+def plot(args, result_dir):
```
I would add markers to show which part of the line were actually measured

lanice replied:
Good point :+1:

lanice commented on src/benchmarklib/benchmark_utils.hpp:
```diff
@@ -113,6 +110,8 @@ struct BenchmarkConfig {
   const std::optional<std::string> output_file_path = std::nullopt;
   const bool enable_scheduler = false;
+  const uint cores = 0;
+  const uint clients = 1;
```
The reason behind this is that, if you do specify `--scheduler` but not `--clients`, you basically run the "normal" benchmarks with active scheduler, as we had before when we just added `--scheduler` to the benchmark, executing one benchmark after another, but with active scheduler (and now customizable cores). Also, if the default is != 1, then we'd have to check during setup if the scheduler is active, and set clients back to 1 for the "normal" benchmarks.
Which is fine by me, let me know what you prefer.

lanice commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -103,40 +104,106 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
+  auto finished_query_set_runs = std::atomic_uint{0};
+  auto finished_queries_total = std::atomic_uint{0};
+  // For the iteration durations vector, we need a mutex
+  auto result_mutex = std::mutex{};
```
You mean making the `num_iterations` and `duration` atomic and `iteration_durations` into a concurrent vector? But we wouldn't need to pre-allocate elements then, right?
```
// benchmark_utils.hpp

struct QueryBenchmarkResult {
  size_t num_iterations = 0;
  Duration duration = Duration{};
  std::vector<Duration> iteration_durations;
};
```

mrks replied:
> But we wouldn't need to pre-allocate elements then

We wouldn't have to, but the default growth algorithm means that we get a lot of fragmentation in the beginning if we don't

lanice replied:
Okay, so for some reason I can not get it to work so far, might need some help. When I change just one member (`num_iterations`) of the struct to atomic:
```
struct QueryBenchmarkResult {
  std::atomic<size_t> num_iterations = 0;
  Duration duration = Duration{};
  std::vector<Duration> iteration_durations;
};
```
I get the following error when emplacing/inserting into the result map (`std::unordered_map<std::string, QueryBenchmarkResult>`):
```
...
/home/lanice/develop/epic/opossumdb/src/benchmarklib/benchmark_runner.cpp:105:70:   required from here
/usr/include/c++/7/ext/new_allocator.h:136:4: error: no matching function for call to ‘std::pair<const std::__cxx11::basic_string<char>, opossum::QueryBenchmarkResult>::pair(const std::__cxx11::basic_string<char>&, opossum::QueryBenchmarkResult)’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/7/utility:70:0,
                 from /usr/include/c++/7/algorithm:60,
                 from /home/lanice/develop/epic/opossumdb/third_party/json/json.hpp:32,
                 from /home/lanice/develop/epic/opossumdb/src/benchmarklib/benchmark_runner.cpp:1:
/usr/include/c++/7/bits/stl_pair.h:431:9: note: candidate: template<class ... _Args1, long unsigned int ..._Indexes1, class ... _Args2, long unsigned int ..._Indexes2> std::pair<_T1, _T2>::pair(std::tuple<_Args1 ...>&, std::tuple<_Args2 ...>&, std::_Index_tuple<_Indexes1 ...>, std::_Index_tuple<_Indexes2 ...>)
         pair(tuple<_Args1...>&, tuple<_Args2...>&,
         ^~~~
/usr/include/c++/7/bits/stl_pair.h:431:9: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/7/bits/c++allocator.h:33:0,
                 from /usr/include/c++/7/bits/allocator.h:46,
                 from /usr/include/c++/7/string:41,
                 from /usr/include/c++/7/stdexcept:39,
                 from /usr/include/c++/7/array:39,
                 from /home/lanice/develop/epic/opossumdb/third_party/json/json.hpp:33,
                 from /home/lanice/develop/epic/opossumdb/src/benchmarklib/benchmark_runner.cpp:1:
/usr/include/c++/7/ext/new_allocator.h:136:4: note:   types ‘std::tuple<_Tps ...>’ and ‘const std::__cxx11::basic_string<char>’ have incompatible cv-qualifiers
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
...
```
Caused when constructing the pair in the `emplace` calls in:
```
// Init results
for (const auto& named_query : _queries) {
  const auto& name = named_query.first;
  _query_results_by_query_name.emplace(name, QueryBenchmarkResult{});
}
```
And:
```
auto result = QueryBenchmarkResult{};
result.num_iterations = finished_query_runs;
result.duration = state.benchmark_duration;
result.iteration_durations = iteration_durations;
_query_results_by_query_name.emplace(name, result);
```

mrks replied:
You can't copy an atomic, so you need to find a way (references/moving) where you don't need to.

lanice commented:
Here a sample output plot after the last changes:
![result_plots](https://user-images.githubusercontent.com/1871704/46268857-67cd8800-c53d-11e8-8813-4fb690af9047.png)

lanice commented on scripts/benchmark_multithreaded.py:
```diff
@@ -28,7 +28,7 @@
 # python3 ./scripts/benchmark_multithreaded.py -e ./build-release/hyriseBenchmarkTPCH --runs 1000 -v --scale 1 --cores 10 20 30 --queries 1,3,6,12 --clients 20
 
 
-MAX_CORE_COUNT = multiprocessing.cpu_count()
```
Thanks for your changes! I agree with everything except this one. The Topology already handles this if we are on a non-NUMA system. On a NUMA-system it uses all available CPUs. I don't think we should restrict this in two different spots, if it has to go through one of them anyway. Also this would easily get overridden when specifying --cores or --max-cores, while still possibly being changed by the Topology.

lanice commented on scripts/benchmark_multithreaded.py:
```diff
@@ -0,0 +1,218 @@
+# python3 ./scripts/benchmark_multithreaded.py -e ./build-release/hyriseBenchmarkTPCH --runs 1000 -v --scale 1 --cores 10 20 30 --queries 1,3,6,12 --clients 20
+
+
+MAX_CORE_COUNT = max(1, multiprocessing.cpu_count() - 1)  # Keep one core empty, same as in topology.cpp
```
Thanks for your changes! I agree with everything except this one. The Topology already handles this if we are on a non-NUMA system. On a NUMA-system it uses all available CPUs. I don't think we should restrict this in two different spots, if it has to go through one of them anyway. Also this would easily get overridden when specifying --cores or --max-cores, while still possibly being changed by the Topology.

lanice replied:
Okay, that makes sense. After thinking about it, I honestly think we should remove this limitation completely. If I want to benchmark something on my laptop, with 4 cores, and go `hyriseBenchmarkTPCH --cores 4 ...`, then I want to benchmark it with 4 cores, and not with 3. If I want to leave one core out because I want to work on something else while benchmarking, then I can specify that manually. We don't restrict CPUs on a NUMA system either, if we say use 80 cores, we do it.
Basically I don't want the Benchmark to "not allow" me to fully utilize my system for benchmarking if I want to... right?

lanice replied:
Which btw, funnily, I did the whole time before. Since I installed libnuma some time ago to test something, CMake thinks I'm on a NUMA system, so I never even got that limitation when benchmarking on my laptop, because I always had a "NUMA" Topology with only 1 node and 4 cores. Worked perfectly, no freeze. Of course I didn't do anything else during that benchmark, but that was my intent, because I knew that all 4 cores would be used.
That's why I forgot about that limitation, because it didn't come up when I tested it ^^

mrks replied:
Agreed. Please get rid of it.

lanice commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -103,52 +111,157 @@ void BenchmarkRunner::_benchmark_permuted_query_sets() {
+            finished_query_set_runs++;
+          }
+
+          if (!state.is_done()) { // To prevent queries to add their results after the time is up
```
I added this, because after the time is up we still "wait" for the currently running clients to finish, but of course we don't want them to add their results anymore.
